packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -219,282 +209,70 @@ export class OdspDocumentService implements IDocumentService {
-                this.scheduleJoinSessionRefresh(response.refreshAfterDeltaMs)
-                    .catch((error) => {
-                        const canRetry = canRetryOnError(error);
-                        // Only record error event in case it is non retriable.
-                        if (!canRetry) {
-                            this.mc.logger.sendErrorEvent({
-                                eventName: "JoinSessionRefreshError",
-                                details: JSON.stringify(props),
-                            },
-                            error,
-                            );
-                        }
-                    });
-            } else {
-                // Logging just for informational purposes to help with debugging as this is a new feature.
-                this.mc.logger.sendTelemetryEvent({
-                    eventName: "JoinSessionRefreshNotScheduled",
-                    details: JSON.stringify(props),
-                });
-            }
+        if (this.socketModuleP === undefined) {
+            this.socketModuleP = new Deferred();
+        } else {
+            await this.socketModuleP.promise;
         }
-        return response.joinSessionResponse;
-    }
-
-    private calculateJoinSessionRefreshDelta(responseFetchTime: number, refreshSessionDurationSeconds: number) {
-        // 30 seconds is buffer time to refresh the session.
-        return responseFetchTime + ((refreshSessionDurationSeconds * 1000) - 30000) - Date.now();
+        const deltaStream = await this.getDelayLoadedDeltaStream();
+        return deltaStream.connectToDeltaStream(client);
     }
 
     /**
-     * Creats a connection to the given delta stream endpoint
-     *
-     * @param tenantId - the ID of the tenant
-     * @param documentId - document ID
-     * @param token - authorization token for delta service
-     * @param io - websocket library
-     * @param client - information about the client
-     * @param webSocketUrl - websocket URL
+     * This dynamically imports the module for loading the delta connection. In many cases the delta stream, is not
+     * required during the critical load flow. So this way we don't have to bundle this in the initial bundle and can
+     * import this later on when required.
+     * @returns - delta stream object.
      */
-    private async createDeltaConnection(
-        tenantId: string,
-        documentId: string,
-        token: string | null,
-        io: typeof SocketIOClientStatic,
-        client: IClient,
-        webSocketUrl: string,
-    ): Promise<OdspDocumentDeltaConnection> {
-        const startTime = performance.now();
-        const connection = await OdspDocumentDeltaConnection.create(
-            tenantId,
-            documentId,
-            token,
-            io,
-            client,
-            webSocketUrl,
-            this.mc.logger,
-            60000,
+    private async getDelayLoadedDeltaStream() {
+        if (this.odspDelayLoadedDeltaStream) {
```

Comment:
I believe the assumption of the code (as it's structured right now, with usage of this.socketModuleP) is that we always get here with this.odspDelayLoadedDeltaStream === undefined

Reply:
Yes, this seems just a protection in case we call this somewhere else after it is initialized, then we don't have to go through this flow again. Rather it would just return it. So, we only the below logic to be executed only once.

Reply:
I'd suggest asserting that it's called only once (if that's current design), i.e. built in defenses against incorrect workflows